runtime/eventring.c

```diff
@@ -0,0 +1,543 @@
+static int current_ring_total_size;
+static char_os *eventring_path;
+static char_os *current_ring_loc = NULL;
+
+#ifdef _WIN32
+static HANDLE ring_file_handle;
+static HANDLE ring_handle;
+#endif
+
+/* This comes from OCAMLRUNPARAMs and is initialised in caml_eventring_init */
+static int ring_size_words;
+
+static atomic_uintnat eventring_enabled = 0;
+static atomic_uintnat eventring_paused = 0;
+
+static void write_to_ring(ev_category category, ev_message_type type,
+                          int event_id, int event_length, uint64_t *content,
+                          int word_offset);
+
+void caml_eventring_init() {
+  eventring_path = caml_secure_getenv(T("OCAML_EVENTRING_DIR"));
+
+  ring_size_words = 1 << caml_params->eventring_size;
+
+  if (caml_secure_getenv(T("OCAML_EVENTRING_START"))) {
+    caml_eventring_start();
+  }
+}
+
+/* Remove the ringbuffers. This is only to be called in a stop-the-world */
+static void teardown_eventring(caml_domain_state *domain_state, void *data,
+                               int num_participating,
+                               caml_domain_state **participating_domains) {
+  caml_global_barrier();
+  if (participating_domains[0] == domain_state) {
+#ifdef _WIN32
+    UnmapViewOfFile(current_metadata);
+    CloseHandle(ring_file_handle);
+    CloseHandle(ring_handle);
+#else
+    munmap(current_metadata, current_ring_total_size);
+
+    int remove_file = *(int*)data;
+
+    if( remove_file ) {
+      unlink(current_ring_loc);
+    }
+#endif
+
+    caml_stat_free(current_ring_loc);
+    current_metadata = NULL;
+
+    atomic_store_rel(&eventring_enabled, 0);
+  }
+  caml_global_barrier();
+}
+
+void caml_eventring_post_fork() {
+  /* We are after a call to fork (which can only happen when there is a single
+     domain) and no mutator code that can spawn a new domain can have run yet.
+     Let's be double sure. */
+  CAMLassert(caml_domain_alone());
+
+  if (atomic_load_acq(&eventring_enabled)) {
+    /* unmmap eventrings from the parent but don't remove the files */
+    int remove_file = 0;
+    caml_try_run_on_all_domains(&teardown_eventring, &remove_file, NULL);
+
+    /* We still have the path and ring size from our parent */
+    caml_eventring_start();
+  }
+}
+
+/* Return the current location for the ring buffers of this process. This is
+  used in the consumer to read the ring buffers of the current process */
+char_os* caml_eventring_current_location() {
+  if( atomic_load_acq(&eventring_enabled) ) {
+    return current_ring_loc;
+  } else {
+    return NULL;
+  }
+}
+
+/* Write a lifecycle event and then trigger a stop the world to tear down the
+  ring buffers */
+void caml_eventring_destroy() {
+  if (atomic_load_acq(&eventring_enabled)) {
+    write_to_ring(EV_RUNTIME, EV_LIFECYCLE, EV_RING_STOP, 0, NULL, 0);
+
+    /* clean up eventring when we exit. */
+    int remove_file = 1;
+    caml_try_run_on_all_domains(&teardown_eventring, &remove_file, NULL);
```

Comment:
Again, this call can fail to do the work, it should be in loop `while` `eventring_enabled`.

Note: if two domains decide to destroy the eventring, one will "win" and the `remove_file` setting of the other one will be ignored. This looks okay because `remove_file` is only "false" in just-forked children, which cannot race with `_destroy()`. Maybe there could be a comment about this:

```
/* Note: any concurrent call to disable eventring will also use [remove_file = 1],
   as only [caml_eventring_post_fork] uses [remove_file = 0] and that call cannot
   be executed in parallel with another. In particular, the file will be removed no
   matter which of the concurrent STW sections win to run [teardown_eventring] first. */
```

Reply:
Thanks. This has been fixed in https://github.com/ocaml/ocaml/pull/10964/commits/339ef2ec4d40437dc390eeded87d2a787f73c404 and https://github.com/ocaml/ocaml/pull/10964/commits/92b9882fcd187bd8873fc5901748f448f7e198d6 and I've adopted your comment.